examples/jacobi/jacobi.cpp

```diff
@@ -0,0 +1,472 @@
+/* charmlite demo
+ *
+ * author: aditya bhosale <adityapb@illinois.edu>
+ *
+ * based on jacobi2d implementation in charm++
+ * https://github.com/UIUC-PPL/charm/tree/main/examples/charm%2B%2B/jacobi2d-2d-decomposition
+ */
+
+#include <cstring>
+#include <charmlite/charmlite.hpp>
+
+#define LEFT 0
+#define RIGHT 1
+#define TOP 2
+#define BOTTOM 3
+
+#define THRESHOLD 0.0004
+#define MAX_ITER 1000
+
+class index2d
+{
+public:
+    int x, y;
+
+    index2d(int x_=0, int y_=0) :
+        x(x_), y(y_)
+    {
+    }
+
+    void operator+=(const index2d& idx)
+    {
+        x += idx.x;
+        y += idx.y;
+    }
+
+    bool operator==(const index2d& idx)
+    {
+        return idx.x == x && idx.y == y;
+    }
+
+    bool operator!=(const index2d& idx)
+    {
+        return idx.x != x || idx.y != y;
+    }
+};
+
+template <>
+struct cmk::index_view<index2d>
+{
+    static index2d& decode(chare_index_t& idx)
+    {
+        return reinterpret_cast<index2d&>(idx);
+    }
+
+    static const index2d& decode(const chare_index_t& idx)
+    {
+        return reinterpret_cast<const index2d&>(idx);
+    }
+
+    static chare_index_t encode(const index2d& idx)
+    {
+        chare_index_t encoded_idx = 0;
+        std::memcpy(&encoded_idx, &idx, sizeof(index2d));
+        return encoded_idx;
+    }
+};
+
+struct ghost_message : public cmk::message
+{
+    int direction;
+    int size;
+    char* data;
+
+    ghost_message(int direction_, int size_)
+        : cmk::message(cmk::message_helper_<ghost_message>::kind_,
+                sizeof(message) + 2 * sizeof(int) + sizeof(char*) + size_ * sizeof(double))
+          , direction(direction_)
+          , size(size_)
+          , data((char*) this + 2 * sizeof(int) + sizeof(char*) + sizeof(message))
+    {
+    }
+
+    static void pack(cmk::message_ptr<ghost_message>& msg)
+    {
+        msg->data = msg->data - (std::uintptr_t) msg.get();
+    }
+
+    static void unpack(cmk::message_ptr<ghost_message>& msg)
+    {
+        msg->data = msg->data + (std::uintptr_t) msg.get();
+    }
+};
+
+struct completion_message : public cmk::plain_message<completion_message>
```

Comment:
I would recommend using a `data_message` here, then contributing `logical_and` to _reduction.hpp_ in the style of add.